torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1976,6 +1976,119 @@ def print_results(grouped, indent, result_fn):
+    # the captured FX nodes from the old ShapeEnv by corresponding ones from
+    # shape_env.
+    def __call__(
+        self,
+        shape_env: Optional["ShapeEnv"] = None,
+        mapping: Optional[Callable[[torch.fx.Node], torch.fx.Node]] = None
+    ) -> Any:
+        # Special handling for the constructor event.
+        if self.f == ShapeEnv:
+            assert shape_env is None and self.args is None and self.kwargs is not None
+            return ShapeEnv(**self.kwargs)
+
+        assert shape_env is not None
+
+        args = list(self.args or list())
+        kwargs = dict(self.kwargs or dict())
+
+        # Converts FX nodes using the mapping argument.
+        def convert_node(x: Any) -> Any:
+            return x if mapping is None or not isinstance(x, torch.fx.Node) else mapping(x)
+
+        # Replaces the value of an specific argument by the result of fn.
+        def replacearg(index: int, key: str, fn: Callable):
+            if index < len(args):
+                args[index] = fn(args[index])
+            if key in kwargs:
+                kwargs[key] = fn(kwargs[key])
+
+        if self.is_create_fx_call_function():
+            # ShapeEnv.create_fx_call_function:
+            # "args" parameter is a tuple of FX nodes from the FX graph of the old ShapeEnv.
+            # They must be replaced, since a "call_function" FX node with this tuple as argument
+            # will be added to the FX graph of the new shape_env.
+            replacearg(index=1, key="args", fn=lambda args: tuple(convert_node(a) for a in args))
+        if self.is_evaluate_expr():
+            # ShapeEnv.evaluate_expr:
+            # "fx_node" parameter is an (optional) FX node that represents the evaluate expression.
+            # They must be replaced, since it will be part of a "call_function" FX node for
+            # torch._assert, which will be added to the FX graph of the new shape_env.
+            replacearg(index=3, key="fx_node", fn=convert_node)
+
+        # Actually call the method with the converted arguments.
+        return self.f(shape_env, *args, **kwargs)
+
+    def __str__(self) -> str:
+        name = self.name if self.name is not None else self.f.__name__
+        return f"event: {name} ({self.args}, {self.kwargs})"
+
+    def is_create_fx_call_function(self) -> bool:
+        return self.name == "create_fx_call_function"
+
+    def is_evaluate_expr(self) -> bool:
+        return self.name == "evaluate_expr"
+
+# Decorator for recording the given function as a replayable event.
+#
+# This decorator should be used at every function that mutates the state of
+# ShapeEnv in some way that affects the resulting issued guards (i.e. when
+# ShapeEnv.produce_guards is called).
+#
+# remember_fakes_length: flags whether we should add the length of the
+# ShapeEnv.tracked_fakes list to the event. This is used for calling
+# ShapeEnv.produce_guards at an arbitrary time (assuming the fakes list only
+# increases monotonically).
```

Comment:
If I add a new method which mutates the state of ShapeEnv, but I didn't read this code carefully enough, how will I know to apply this decorator? What test will tell me I did it wrong?